From 44d96f48f8f95a105480febf530bf9e43f8cc149 Mon Sep 17 00:00:00 2001
From: Natanael Copa <ncopa@alpinelinux.org>
Date: Wed, 6 Dec 2023 17:43:47 +0100
Subject: [PATCH] nocloud: support set data source via smbios in qemu

With this we can set the data source with qemu args:
  -smbios type=1,serial=ds='nocloud;s=http://10.0.2.2:8000/'

and tiny-cloud would fetch https://10.0.2.2:8000/{meta,user}-data
---
 lib/tiny-cloud/cloud/nocloud/imds |  6 +++---
 tests/imds.test                   | 14 +++++++++++++-
 2 files changed, 16 insertions(+), 4 deletions(-)

diff --git a/lib/tiny-cloud/cloud/nocloud/imds b/lib/tiny-cloud/cloud/nocloud/imds
index 30b115c..8f70522 100644
--- a/lib/tiny-cloud/cloud/nocloud/imds
+++ b/lib/tiny-cloud/cloud/nocloud/imds
@@ -7,9 +7,9 @@ NOCLOUD_FILES="meta-data user-data vendor-data network-config"
 is_nocloud_loaded() { [ -f "$TINY_CLOUD_VAR/.nocloud_loaded" ]; }
 
 _load_nocloud_cmdline() {
-	local kopt kv k v data
-
-	for kopt in $(cat "$ROOT/proc/cmdline" 2>/dev/null); do
+	local kopt kv k v data smbios
+	for kopt in $(cat "$ROOT/proc/cmdline" 2>/dev/null) \
+			$(grep '^ds=nocloud' "$ROOT"/sys/class/dmi/id/product_serial 2>/dev/null) ; do
 		echo "$kopt" | grep -qE '(^|=)ds=nocloud(-net)?;' || continue
 		for kv in $(echo "${kopt#*;}" | tr \; ' '); do
 			k=$(echo "$kv" | cut -d= -f1)
diff --git a/tests/imds.test b/tests/imds.test
index 025ab10..ec98b6d 100755
--- a/tests/imds.test
+++ b/tests/imds.test
@@ -32,7 +32,8 @@ init_tests \
 	imds_ssh_keys_nocloud \
 	imds_ssh_keys_oci \
 	\
-	imds_nocloud_cmdline_local_hostname
+	imds_nocloud_cmdline_local_hostname \
+	imds_nocloud_smbios_local_hostname
 
 imds_help_body() {
 	atf_check -o match:"Usage: imds" imds -h
@@ -125,3 +126,14 @@ imds_nocloud_cmdline_local_hostname_body() {
 			imds @local-hostname
 	done
 }
+
+imds_nocloud_smbios_local_hostname_body() {
+	atf_require_prog yx
+	mkdir -p sys/class/dmi/id
+	for key in h local-hostname; do
+		echo "ds=nocloud;$key=myhostname" > sys/class/dmi/id/product_serial
+		CLOUD=nocloud atf_check \
+			-o match:'^myhostname$' \
+			imds @local-hostname
+	done
+}
-- 
2.43.0

